quote-book/src/synchronized_quote_book.rs

```diff
@@ -90,6 +163,156 @@ where
+                                    quote.id(),
+                                    err
+                                );
+                            }
+                        }
+                    }
+                    Err(err) => {
+                        log::error!(
+                            self.logger,
+                            "Failed to sync to block_index {}: {:?}",
+                            last_processed_block_index,
+                            err
+                        );
+                    }
+                }
+                let mut shared_state = self.shared_state.lock().expect("mutex poisoned");
+                // this is next_block_index because next_block_index is actually the block
+                // we just processed
+                shared_state.highest_processed_block_index = last_processed_block_index;
+            }
+            std::thread::sleep(Self::POLLING_FREQUENCY);
+        }
+    }
+
+    /// Attempt to load the next block that we
+    /// are aware of and remove quotes that match key images from it.
+    /// Returns true if we might have more block data to load.
+    fn load_block_data(&mut self) -> bool {
+        // Default to true: if there is an error, we may have more work, we don't know
+        let mut may_have_more_work = true;
+
+        match self.db.get_block_contents(self.next_block_index) {
+            Err(LedgerError::NotFound) => may_have_more_work = false,
+            Err(e) => {
+                log::error!(
+                    self.logger,
+                    "Unexpected error when checking for block data {}: {:?}",
+                    self.next_block_index,
+                    e
+                );
+                std::thread::sleep(Self::ERROR_RETRY_FREQUENCY);
+            }
+            Ok(block_contents) => {
+                // Filter keyimages in quotebook
+                for key_image in block_contents.key_images {
+                    match self.quotebook.remove_quotes_by_key_image(&key_image) {
+                        Ok(quotes) => {
+                            for quote in quotes {
+                                log::info!(self.logger, "Quote {} removed", quote.id());
+                                if let Err(err) = self
+                                    .msg_bus_tx
+                                    .blocking_send(Msg::SciQuoteRemoved(*quote.id()))
+                                {
+                                    log::error!(
+                                        self.logger,
+                                        "Failed to send SCI quote {} removed message to message bus: {:?}",
+                                        quote.id(),
+                                        err
+                                    );
+                                }
+                            }
+                        }
+                        Err(err) => {
+                            log::error!(
```

Comment:
this also feels a little squirrelly  -- if we can't remove a key image from the quotebook, we just log it, move on and forget about it?

Reply:
i want to understand more about how this could fail -- for the in-memory one it seems that it can't, but maybe for a sql-backed one it can?

it might be correct to retry this in a loop until we remove it without getting an error.